Use StringView more in ScriptElement and EmailInputType

annevk opened the pull request:
#### f8eaba1e91892d09b40add74c79e92bff1a8fc49
<pre>
Use StringView more in ScriptElement and EmailInputType
<a href="https://bugs.webkit.org/show_bug.cgi?id=257500">https://bugs.webkit.org/show_bug.cgi?id=257500</a>
rdar://110047074

Reviewed by Darin Adler.

Move the logic for the for and event attributes to HTMLScriptElement as
SVGScriptElement does not have to care for them and also avoid going
from AtomString to String and then allocating a new String when
trimming.

As RegularExpression::match() already takes a StringView make
isValidEmailAddress() match that.

* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::isScriptForEventSupported const): Deleted.
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::isScriptPreventedByForAndEventAttributes const):
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::isValidEmailAddress):
(WebCore::EmailInputType::typeMismatchFor const):
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::isScriptPreventedByForAndEventAttributes const):
(WebCore::HTMLScriptElement::forAttributeValue const): Deleted.
(WebCore::HTMLScriptElement::eventAttributeValue const): Deleted.
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/svg/SVGScriptElement.h:

Canonical link: <a href="https://commits.webkit.org/264736@main">https://commits.webkit.org/264736@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de72495728b8a4b2f0b58424a75c2ebef7008220

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8478 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8771 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8990 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10145 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8524 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8488 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8718 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11393 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8624 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9667 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10300 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6965 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7758 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15309 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8086 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7905 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11260 "6 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8377 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7663 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2055 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11874 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8122 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/ScriptElement.h:
```diff
@@ -119,11 +118,11 @@ class ScriptElement {
     virtual ReferrerPolicy referrerPolicy() const = 0;
     virtual RequestPriority fetchPriorityHint() const { return RequestPriority::Auto; }
 
+    virtual bool isScriptPreventedByForAndEventAttributes() const { return false; };
```
I think it’s unnecessary to actually mention `ForAndEvent` in this function name and it makes the name seem inelegant to me.

darinadler commented on Source/WebCore/html/EmailInputType.cpp:
```diff
@@ -41,7 +41,7 @@ using namespace HTMLNames;
 static constexpr ASCIILiteral emailPattern = "^[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$"_s;
 
-static bool isValidEmailAddress(const String& address)
+static bool isValidEmailAddress(const StringView address)
```
Should just be `StringView`, not `const StringView`.

darinadler commented on Source/WebCore/html/EmailInputType.cpp:
```diff
@@ -66,9 +66,9 @@ bool EmailInputType::typeMismatchFor(const String& value) const
         return false;
     if (!element()->multiple())
-        return !isValidEmailAddress(value);
+        return !isValidEmailAddress(StringView(value));
```
Passing a `String` to a `StringView` argument should convert without explicit conversion.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -183,6 +173,21 @@ void HTMLScriptElement::dispatchLoadEvent()
 
+bool HTMLScriptElement::isScriptPreventedByForAndEventAttributes() const
+{
+    const auto eventAttribute = attributeWithoutSynchronization(eventAttr);
```
Should be `auto&` instead of `const auto` to avoid unnecessary reference count churn.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -183,6 +173,21 @@ void HTMLScriptElement::dispatchLoadEvent()
+bool HTMLScriptElement::isScriptPreventedByForAndEventAttributes() const
+{
+    const auto eventAttribute = attributeWithoutSynchronization(eventAttr);
+    const auto forAttribute = attributeWithoutSynchronization(forAttr);
```
Should be `auto&` instead of `const auto` to avoid unnecessary reference count churn.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -183,6 +173,21 @@ void HTMLScriptElement::dispatchLoadEvent()
+        if (!equalLettersIgnoringASCIICase(StringView(forAttribute).trim(isASCIIWhitespace<UChar>), "window"_s))
+            return true;
+
+        const auto eventAttributeView = StringView(eventAttribute).trim(isASCIIWhitespace<UChar>);
```
Just `auto` here would be better. Any local could be `const` if it’s not reassigned but we normally omit that for brevity.

Surprised that this rule applies only if the `for` attribute is present, but regardless of its value.

annevk replied:
Dropped `const`.

Yeah, see step 20 of https://html.spec.whatwg.org/multipage/scripting.html#prepare-the-script-element. Back in the day Internet Explorer probably did something more interesting with these attributes.

darinadler commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -183,6 +173,21 @@ void HTMLScriptElement::dispatchLoadEvent()
+    const auto eventAttribute = attributeWithoutSynchronization(eventAttr);
+    const auto forAttribute = attributeWithoutSynchronization(forAttr);
+    if (!eventAttribute.isNull() && !forAttribute.isNull()) {
+        if (!equalLettersIgnoringASCIICase(StringView(forAttribute).trim(isASCIIWhitespace<UChar>), "window"_s))
```
Surprised that this rule applies only if the `event` attribute is present but regardless of its value.

annevk committed:
Use StringView more in ScriptElement and EmailInputType
https://bugs.webkit.org/show_bug.cgi?id=257500
rdar://110047074

Reviewed by Darin Adler.

Move the logic for the for and event attributes to HTMLScriptElement as
SVGScriptElement does not have to care for them and also avoid going
from AtomString to String and then allocating a new String when
trimming.

As RegularExpression::match() already takes a StringView make
isValidEmailAddress() match that.

* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::isScriptForEventSupported const): Deleted.
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::isScriptPreventedByForAndEventAttributes const):
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::isValidEmailAddress):
(WebCore::EmailInputType::typeMismatchFor const):
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::isScriptPreventedByForAndEventAttributes const):
(WebCore::HTMLScriptElement::forAttributeValue const): Deleted.
(WebCore::HTMLScriptElement::eventAttributeValue const): Deleted.
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/svg/SVGScriptElement.h:

Canonical link: https://commits.webkit.org/264736@main
